scintillometry/fourier/base.py

```diff
@@ -0,0 +1,290 @@
+            If `None` (default), uses an unscaled forward transform and 1 / n
+            scaled inverse transform.  If 'ortho', uses a 1 / sqrt(n) scaling
+            for both.
+        verify : bool, optional
+            Verify setup is successful and self-consistent.
+        """
+        # Extract information if user passed in a dummy array.
+        if isinstance(time_data, np.ndarray):
+            time_data = {'shape': time_data.shape,
+                         'dtype': time_data.dtype.name}
+
+        # Set axes and normalization.  If axes is None, cycle through all axes
+        # (like with numpy.fft).
+        if axes is None:
+            axes = tuple(range(len(time_data['shape'])))
+        else:
+            # If axes is an integer, convert to a 1-element tuple.
+            try:
+                axes = operator.index(axes)
+            # If not, typecast to a tuple.
+            except TypeError:
+                axes = tuple(axes)
+            else:
+                axes = (axes,)
+        self._axes = axes
+        self._norm = norm if norm == 'ortho' else None
+
+        # Determine frequency-domain shape.
+        freq_shape = list(time_data['shape'])
+        # If time-domain data is real, halve the relevant axis in frequency.
+        if 'float' in time_data['dtype']:
+            freq_shape[self.axes[-1]] = (
+                freq_shape[self.axes[-1]] // 2 + 1)
+
+        # Store time and frequency-domain array shapes.
+        self._data_format = {'time_shape': time_data['shape'],
+                             'time_dtype': time_data['dtype'],
+                             'freq_shape': tuple(freq_shape),
+                             'freq_dtype': freq_dtype}
+
+        if verify:
+            self.verify()
+
+    def verify(self):
+        """Verify setup is successful and self-consistent."""
+        assert 'complex' in self.data_format['freq_dtype'], (
+            "frequency-domain data must be complex.")
+        assert len(self.axes) > 0, "must transform over one or more axes!"
+
+    @property
+    def axes(self):
+        """Axes over which to perform the FFT."""
+        return self._axes
+
+    @property
+    def norm(self):
+        """Normalization convention.
+
+        As in `numpy.fft`, `None` is an unscaled forward transform and 1 / n
+        scaled inverse one, and 'ortho' is a 1 / sqrt(n) scaling for both.
+        """
+        return self._norm
+
+    @property
+    def data_format(self):
+        """Shapes and dtypes of the FFT arrays.
+
+        'time_' and 'freq_' entries are for time and frequency-domain arrays,
+        respectively.
+        """
+        return self._data_format
+
+    def fft(self, a):
+        """Placeholder for forward FFT."""
+        raise NotImplementedError()
+
+    def ifft(self, a):
+        """Placeholder for inverse FFT."""
+        raise NotImplementedError()
+
+    def fftfreq(self, a_length, sample_rate=None, positive_freqs_only=False):
```

Comment:
I think `sample_rate` should come from `__init__` (fine to allow it to be `1.`) - similarly, why `a_length` when we know the size? I don't think this would need any arguments.